# Dynamic Time Wrapping

Dynamic Time Warping (DTW) adalah sebuah algoritma untuk mengukur *kemiripan* antara dua deret waktu (*time series*)
yang mungkin memiliki panjang berbeda atau kecepatan berbeda.  
DTW bekerja dengan mencari *alignment* terbaik antara dua sinyal sehingga total jarak antara titik-titik yang dibandingkan menjadi minimum.

DTW sering dipakai pada:
- Speech recognition  
- Analisis sinyal sensor  
- Motion tracking  
- Klasifikasi dan clustering time series  

## Library yang dibutuhkan

In [41]:
import librosa
import os
import pandas as pd
import numpy as np

## Mengekstrak data time series dari data seara dengan label 'Buka' dan 'tutup'

In [42]:
folder = '/content/drive/MyDrive/data_PSD/data_suara_time_wrappping'

data_time_series = []
for i in os.listdir(folder):
  if i.endswith(('.wav', '.mp3')):
    path_file = os.path.join(folder, i)
    x, sr = librosa.load(path_file, sr = 4000)
    data_time_series.append(x)


## Hasil 10 data time series yang dihasilkan

In [43]:
print("10 data pertama data time series suara")
for j in range (10):
  print(f"\n indeks ke- {j}, data = {data_time_series[j]}")


10 data pertama data time series suara

 indeks ke- 0, data = [6.4028427e-10 1.4551915e-10 4.0745363e-10 ... 7.3487044e-04 8.2284957e-04
 9.0314628e-04]

 indeks ke- 1, data = [ 5.8207661e-11  8.7311491e-11 -2.9103830e-11 ...  2.1789318e-04
  3.8899285e-05 -4.7112429e-05]

 indeks ke- 2, data = [-2.1827873e-11  0.0000000e+00 -7.2759576e-12 ... -3.8544822e-04
 -3.1474186e-04 -4.0618455e-04]

 indeks ke- 3, data = [ 1.4551915e-11 -2.1827873e-11  7.2759576e-12 ... -6.0990907e-04
 -5.1070319e-04 -7.4119045e-04]

 indeks ke- 4, data = [ 1.4551915e-11  4.3655746e-11  0.0000000e+00 ... -4.8730351e-04
 -4.5495119e-04 -3.9114399e-04]

 indeks ke- 5, data = [-3.6379788e-11 -1.4551915e-11 -1.4551915e-11 ...  3.2236252e-04
  3.9786194e-04  5.8360019e-04]

 indeks ke- 6, data = [ 0.0000000e+00  3.6379788e-11  1.4551915e-11 ...  3.7378886e-06
 -9.0922724e-05 -1.5665835e-04]

 indeks ke- 7, data = [-5.0931703e-11 -1.4551915e-11  5.8207661e-11 ...  1.8611464e-04
  3.1201294e-04  5.5205508e-04]

 indek

## Data Test

In [45]:
file_y = "/content/drive/MyDrive/data_PSD/test.mp3"

y, sr = librosa.load(file_y, sr = 4000)

print("data time series y = ", y)
print(len(y))

data time series y =  [-1.1641532e-10  5.8207661e-11  0.0000000e+00 ...  5.3613132e-04
  6.1152177e-04  7.7914959e-04]
8736


## Perhitungan Dynamic Time Wrapping

In [46]:
matrix_final = []
for i in range(len(data_time_series)):
  print("Proses Iterasi ke-", i)
  # data time series x
  x = data_time_series[i]

  # panjang setiap data time series
  panjang_x = len(data_time_series[i])
  panjang_y = len(y)

  n = panjang_x + 1
  m = panjang_y + 1

  # inisialisasi matrix akumulasi
  matrix_akumulasi = np.full((n, m), np.inf)
  matrix_akumulasi[0, 0] = 0

  # mengisi matrix akumulasi
  for l in range(panjang_x):
    for m in range(panjang_y):
      matrix_cost = (x[l] - y[m]) ** 2
      matrix_akumulasi[l + 1, m + 1] = matrix_cost + min(matrix_akumulasi[l, m],
                                           matrix_akumulasi[l, m + 1],
                                           matrix_akumulasi[l + 1, m])

  nilai_dwt = np.sqrt(matrix_akumulasi[panjang_x, panjang_y])
  matrix_final.append([i, nilai_dwt])
  print(f"proses iterasi ke- {i} selesai")
print(matrix_final)

Proses Iterasi ke- 0
proses iterasi ke- 0 selesai
Proses Iterasi ke- 1
proses iterasi ke- 1 selesai
Proses Iterasi ke- 2
proses iterasi ke- 2 selesai
Proses Iterasi ke- 3
proses iterasi ke- 3 selesai
Proses Iterasi ke- 4
proses iterasi ke- 4 selesai
Proses Iterasi ke- 5
proses iterasi ke- 5 selesai
Proses Iterasi ke- 6
proses iterasi ke- 6 selesai
Proses Iterasi ke- 7
proses iterasi ke- 7 selesai
Proses Iterasi ke- 8
proses iterasi ke- 8 selesai
Proses Iterasi ke- 9
proses iterasi ke- 9 selesai
Proses Iterasi ke- 10
proses iterasi ke- 10 selesai
Proses Iterasi ke- 11
proses iterasi ke- 11 selesai
Proses Iterasi ke- 12
proses iterasi ke- 12 selesai
Proses Iterasi ke- 13
proses iterasi ke- 13 selesai
Proses Iterasi ke- 14
proses iterasi ke- 14 selesai
Proses Iterasi ke- 15
proses iterasi ke- 15 selesai
Proses Iterasi ke- 16
proses iterasi ke- 16 selesai
Proses Iterasi ke- 17
proses iterasi ke- 17 selesai
Proses Iterasi ke- 18
proses iterasi ke- 18 selesai
Proses Iterasi ke- 19
proses ite

## Menambahkan label berdasarkan real dataset

In [50]:
for i in range(200):
  if i <= 99:
    matrix_final[i].append('buka')
  else:
    matrix_final[i].append('tutup')

## Mengurutkan Nilai DWT

In [54]:
data_urut = sorted(matrix_final, key=lambda x: x[1])
print(data_urut)

[[1, np.float64(2.285439958765221), 'buka'], [0, np.float64(2.3277324383474234), 'buka'], [5, np.float64(2.6836438503501716), 'buka'], [6, np.float64(2.8886372751105642), 'buka'], [2, np.float64(2.9208939969009693), 'buka'], [7, np.float64(3.1307778532462254), 'buka'], [3, np.float64(3.2182455901021054), 'buka'], [4, np.float64(3.31206854135007), 'buka'], [8, np.float64(3.312537657880684), 'buka'], [9, np.float64(3.5183104596360857), 'buka'], [57, np.float64(3.5222394824931214), 'buka'], [31, np.float64(3.553284647464119), 'buka'], [52, np.float64(3.579890368029975), 'buka'], [53, np.float64(3.588033153300932), 'buka'], [70, np.float64(3.6678932072485892), 'buka'], [10, np.float64(3.689102454367805), 'buka'], [50, np.float64(3.6917962561284305), 'buka'], [35, np.float64(3.7048448901163273), 'buka'], [43, np.float64(3.718823971220277), 'buka'], [193, np.float64(3.7528677701114908), 'tutup'], [68, np.float64(3.7659202778434095), 'buka'], [34, np.float64(3.782676446313269), 'buka'], [20, 

## Kesimpulan

### Label Data Test Berdasarkan Nilai DWT Terendah

In [56]:
print(data_urut[0])

[1, np.float64(2.285439958765221), 'buka']


Berdasarkan hal tersebut dapat disimpulkan bahwa data test memiliki label buka, dengan nilai DWT terhadap data index 1 yakni sebesar 2.285439958765221.